### PR TITLE
Add precomputing of normal responses (only 1 model loaded at a time).

### LINKED ISSUE #137: Rewrite code to pre-generate normal responses for normal loss computation

### PR DESCRIPTION
**Pull request by Willmish**
As per #137, this PR introduces a potential major gpu memory saving benefit, by loading a single reference model, precomputing the "pretrained/reference model" loss components of the normal loss for ALL batches, and unlearning on the same model.

This is increases the RAM usage of the unlearning (depending on unlearning sample size), but requires only a single model loaded to memory

Closes #137

### CODE REVIEW

File: llm_unlearn_ucl/unlearn_harm.py [hunk 1]
```diff
@@ -268,23 +268,27 @@ def main(args) -> None:
...
+    for loader in train_normal_loaders:
+        pretrained_model_precomputed_normal_outputs_aggregated.append([])
+        for batch in loader:
+            with torch.no_grad():
+                pretrained_outputs = model(
+                    batch["input_ids"].to(device),
+                    attention_mask=batch["attention_mask"].to(device),
+                    labels=batch["labels"].to(device),
+                )
+                prob_p = torch.nn.functional.softmax(pretrained_outputs.logits, -1)
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:284]: Please consider moving prob_p to `cpu` to save GPU memory.

**Review by TheRootOf3 — CHANGES_REQUESTED** [review]
minor fix pls, other than that looks good champ!

File: llm_unlearn_ucl/unlearn_harm.py (hunk 1 above)
Comment (Willmish) [reply llm_unlearn_ucl/unlearn_harm.py:284]: good spot, see next commit

**Review by TheRootOf3 — APPROVED** [review]
LGTM, thanks! 💝
